File: art2_projects/org.kermeta.art2.core/src/main/java/org/kermeta/art2/core/impl/Art2CoreBean.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package org.kermeta.art2.core.impl;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import org.kermeta.art2.Art2Factory;
import org.kermeta.art2.Art2Package;
import org.kermeta.art2.ContainerRoot;
import org.kermeta.art2.api.service.adaptation.deploy.Art2AdaptationDeployService;
import org.kermeta.art2.api.service.core.handler.Art2ModelHandlerService;
import org.kermeta.art2.api.service.core.kompare.ModelKompareService;
import org.kermeta.art2.framework.Art2XmiHelper;
import org.kermeta.art2adaptation.AdaptationModel;
import org.osgi.framework.Bundle;
import org.osgi.framework.BundleContext;

/**
 *
 * @author ffouquet
 */
public class Art2CoreBean implements Art2ModelHandlerService {

    private ContainerRoot model;
    private List<ContainerRoot> previousModel = new ArrayList();

    @Override
    public ContainerRoot getLastModel() {
        return model;
    }

    @Override
    public void updateModel(ContainerRoot _model) {
        if (_model == null) {
            System.out.println("Null Model");
            return;
        }

        AdaptationModel adaptationModel = kompareService.kompare(model, _model, nodeName);
        /* TODO CALL DEPLOY */
        Boolean deployResult = deployService.deploy(adaptationModel);
        /* TODO CALL ROLLBACK IF NECESSARY */


        /* Transactional TODO */
        previousModel.add(model);
        model = _model;
        /* TODO CALL KOMPARE */



        System.out.println("Deploy result " + deployResult);
    }

    @Override
    public List<ContainerRoot> getPreviousModel() {
        return previousModel;
    }

    public void init() {

        try {
            System.out.println("ART2Node=" + nodeName);

            

            /* DEBUG MODE ONLY */
            ContainerRoot umodel = Art2XmiHelper.load("/Users/ffouquet/Desktop/distart.xmi");
            model = Art2Factory.eINSTANCE.createContainerRoot();
            updateModel(umodel);
            System.out.println("Model DEBUG UPDATED");
            if (true) {
                return;
            }
            System.out.println("NoUse STATMENT");

            File lastModelssaved = this.bundle.getBundleContext().getDataFile("lastModel.xmi");
            if (lastModelssaved.getTotalSpace() != 0) {
                /* Load previous state */
                model = Art2XmiHelper.load(lastModelssaved.getAbsolutePath());
            } else {
                /* INIT EMPTY model */
                model = Art2Factory.eINSTANCE.createContainerRoot();
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void stop() {
        Art2XmiHelper.save(this.bundle.getBundleContext().getDataFile("lastModel.xmi").getAbsolutePath(), model);
    }
    private Bundle bundle = null;
    private BundleContext bundleContext = null;
    private org.kermeta.art2.api.service.core.kompare.ModelKompareService kompareService;
    private org.kermeta.art2.api.service.adaptation.deploy.Art2AdaptationDeployService deployService;
    private String nodeName;

    public void setNodeName(String nodeName) {
        this.nodeName = nodeName;
    }

    public void setDeployService(Art2AdaptationDeployService deployService) {
        this.deployService = deployService;
    }

    public void setKompareService(ModelKompareService kompareService) {
        this.kompareService = kompareService;
    }

    public void setBundle(Bundle bundle) {
        this.bundle = bundle;
    }

    public void setBundleContext(BundleContext bundleContext) {
        this.bundleContext = bundleContext;
    }
}

File: art2_projects/org.kermeta.art2.core/src/main/scala/org/kermeta/art2/core/impl/Art2CoreBean.scala
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package org.kermeta.art2.core.impl

import org.kermeta.art2.Art2Factory
import org.kermeta.art2.ContainerRoot
import org.kermeta.art2.api.service.adaptation.deploy.Art2AdaptationDeployService
import org.kermeta.art2.api.service.core.handler.Art2ModelHandlerService
import org.kermeta.art2.api.service.core.kompare.ModelKompareService
import org.kermeta.art2.framework.Art2Actor
import org.kermeta.art2.framework.Art2PlatformHelper
import org.kermeta.art2.framework.Art2XmiHelper
import org.osgi.framework.BundleContext
import org.slf4j.LoggerFactory
import scala.reflect.BeanProperty
import org.kermeta.art2.framework.merger.Art2PlatformMerger
import org.kermeta.art2.framework.message.Art2PlatformModelUpdate
import org.kermeta.art2.framework.message._
import scala.actors.Actor
import scala.collection.JavaConversions._

class Art2CoreBean extends Art2ModelHandlerService with Art2Actor {

  @BeanProperty var bundleContext : BundleContext = null;
  @BeanProperty var kompareService :org.kermeta.art2.api.service.core.kompare.ModelKompareService = null
  @BeanProperty var deployService :org.kermeta.art2.api.service.adaptation.deploy.Art2AdaptationDeployService = null
  @BeanProperty var nodeName : String = ""

  var models : java.util.ArrayList[ContainerRoot] = new java.util.ArrayList()
  var model : ContainerRoot = Art2Factory.eINSTANCE.createContainerRoot()

  var logger = LoggerFactory.getLogger(this.getClass);

  private def switchToNewModel(c : ContainerRoot)={
    models.add(model)
    model = c
  }

  override def start : Actor={
    logger.info("Start event rec : node name = "+bundleContext.getProperty("art2.node.name"))
    setNodeName(bundleContext.getProperty("art2.node.name"));
    super.start

    var lastModelssaved = bundleContext.getDataFile("lastModel.xmi");
    if (lastModelssaved.length() != 0) {
      /* Load previous state */
      var model = Art2XmiHelper.load(lastModelssaved.getAbsolutePath());
      switchToNewModel(model)
    }
    this
  }

  override def stop() : Unit = {
    Art2XmiHelper.save(bundleContext.getDataFile("lastModel.xmi").getAbsolutePath(), models.head);
    exit
  }

  def act()={
    loop{
      react {
        case updateMsg : Art2PlatformModelUpdate => Art2PlatformHelper.updateNodeLinkProp(model,nodeName, updateMsg.targetNodeName, updateMsg.key, updateMsg.value, updateMsg.networkType, updateMsg.weight)
        case Art2PreviousModel() => reply(models)
        case Art2LastModel() => reply(model) /* TODO DEEP CLONE */
        case Art2UpdateModel(newmodel) => {
            if (newmodel == null) { logger.error("Null model")} else {

              var adaptationModel = kompareService.kompare(model, newmodel, nodeName);
              var deployResult = deployService.deploy(adaptationModel,nodeName);

              if(deployResult){
                //MErge previous model on new model for platform model
                Art2PlatformMerger.merge(newmodel,model)
                switchToNewModel(newmodel)
                logger.info("Deploy result " + deployResult)
              } else {
                //KEEP FAIL MODEL
              }
              reply(deployResult)

            }
          }
        case _ @ unknow=> logger.warn("unknow message  "+unknow)
      }
    }
  }
  
  override def getLastModel : ContainerRoot = (this !? Art2LastModel()).asInstanceOf[ContainerRoot]
  override def updateModel(model : ContainerRoot) : java.lang.Boolean ={ (this !? Art2UpdateModel(model)).asInstanceOf[Boolean] }
  override def getPreviousModel : java.util.List[ContainerRoot] = (this !? Art2PreviousModel).asInstanceOf[java.util.List[ContainerRoot]]

}
